client/client.go

```diff
@@ -249,95 +283,62 @@ func (client *Client) err(e error) {
 	}
 }
 
-func (client *Client) handleResponse(key string, resp *Response) *Response {
-	if h, ok := client.respHandler.get(key); ok {
-		h(resp)
-		return nil
-	}
-	return resp
-}
-
-func (client *Client) handleInner(key string, resp *Response) *Response {
-	if h, ok := client.innerHandler.get(key); ok {
-		h(resp)
-		client.innerHandler.remove(key)
-		return nil
-	}
-	return resp
-}
-
 type handleOrError struct {
 	handle string
 	err    error
 }
 
-func (client *Client) do(funcname string, data []byte, flag rt.PT) (handle string, err error) {
-	if client.conn == nil {
-		return "", ErrLostConn
-	}
-	var result = make(chan handleOrError, 1)
-	client.Lock()
-	defer client.Unlock()
-	client.innerHandler.put("c", func(resp *Response) {
-		if resp.DataType == rt.PT_Error {
-			err = getError(resp.Data)
-			result <- handleOrError{"", err}
-			return
-		}
-		handle = resp.Handle
-		result <- handleOrError{handle, nil}
-	})
-	id := IdGen.Id()
-	req := getJob(id, []byte(funcname), data)
-	req.DataType = flag
-	if err = client.write(req); err != nil {
-		client.innerHandler.remove("c")
-		return
-	}
-	var timer = time.After(client.ResponseTimeout)
-	select {
-	case ret := <-result:
-		return ret.handle, ret.err
-	case <-timer:
-		client.innerHandler.remove("c")
-		return "", ErrLostConn
-	}
-	return
+func (client *Client) request() *request {
+	return client.requestPool.Get().(*request)
+}
+
+func (client *Client) submit(pt rt.PT, funcname string, arg []byte) string {
+	client.outbound <- client.request().submitJob(pt, funcname, IdGen.Id(), arg)
+
+	res := <-client.expected
+
+	client.responsePool.Put(res)
+
+	return res.Handle
 }
 
 // Call the function and get a response.
 // flag can be set to: JobLow, JobNormal and JobHigh
-func (client *Client) Do(funcname string, data []byte,
+func (client *Client) Do(funcname string, arg []byte,
 	flag byte, h ResponseHandler) (handle string, err error) {
-	var datatype rt.PT
+	var pt rt.PT
+
 	switch flag {
 	case rt.JobLow:
-		datatype = rt.PT_SubmitJobLow
+		pt = rt.PT_SubmitJobLow
```

Comment:
Can we spell out pt? What is it supposed to stand for?

Reply:
packettype